---
title: "Tutorial"
page-layout: full
title-block-banner: true
format: 
  html: 
    css: styles/tutorial.css
    toc: true
    toc-depth: 3
    toc-title: "**Table of contents**"
editor: 
  markdown: 
    wrap: 72
---

::: metagenomics
# Downstream processing of fungal metagenomic data {#hidden-section}

![](pictures/fungi.png){width="10%"}

Here, I have illustrated the steps for downstream processing of fungal
metagenomic data using QIIME2 in macOSx.
:::

## Download and install QIIME2

::: download
To download QIIME2 natively, you can follow these steps:

1.  Install [**Miniconda**](https://docs.anaconda.com/miniconda/)
    appropriate to your system.
2.  Then, open terminal window.
3.  Confirm Miniconda is installed. Just write "conda" and press enter.
4.  Now install [**QIIME2 Metagenome
    Distribution**](https://docs.qiime2.org/2024.10/install/native/#install-qiime-2-within-a-conda-environment)
    running following code.

[conda env create -n qiime2-metagenome-2024.10 --file
https://data.qiime2.org/distro/metagenome/qiime2-metagenome-2024.10-py310-osx-conda.yml]{style="color: #0000FF;"}

It will take some time.....
:::

## Activate QIIME2 environment

::: download
1.Run following code to activate QIIME2 environment.

[conda activate qiime2-metagenome-2024.10]{style="color: #0000FF;"}<br>
You can see "qiime" environment being activated
:::

## Importing demultiplexed pair-end sequences

::: download
1.  Make sure your working directory is set to the folder where you have
    saved your demultiplexed sequences.

Run [cd /path/to/your/file]{style="color: #0000FF;"} this will change
your directory

Run [pwd]{style="color: #0000FF;"} check your current working directory

2.  Make sure your sequences file are in .gz and has name format e.g.

**Forward sequences** AF05_SB09_L001_R1_001.fastq.gz

**Reverse sequences** AF05_SB09_L001_R2_001.fastq.gz

Here AF05 and SB09 are the sample information. You can use L001 and 001
for all samples, it will not affect importing demultiplexed sequences.

[qiime tools import Â <br> --type
'SampleData\[PairedEndSequencesWithQuality\]' \\<br> --input-path .
\\<br> --input-format CasavaOneEightSingleLanePerSampleDirFmt \\<br>
--output-path demux-paired-end.qza]{style="color: #0000FF;"}
:::

## Visualize imported paired-end sequences

::: download
To get .qvz file of imported paired-end sequences run following code
<br>

[qiime demux summarize \\<br> --i-data demux-paired-end.qza \\<br>
--o-visualization demux.qzv]{style="color: #0000FF;"}

use [**QIIME2view**](https://view.qiime2.org) to view demux.qzv file<br>
You can see one example in the following figures.
![](pictures/demux_qvz.png){width="100%"}<br> here, you can see the
quality of sequences which will help you to determine the parameters for
denoising sequences for both forward and reverse sequencing
:::

## Denoising

::: download
Denoising step includes cutting out ambiguous bp which are usually
during begining and end of sequence amplification process, followed by
removing chimeric sequences.<br> [qiime dada2 denoise-paired \\<br>
--i-demultiplexed-seqs demux-paired-end.qza \\<br> --p-trim-left-f 15
\\<br> --p-trim-left-r 15 \\<br> --p-trunc-len-f 260 \\<br>
--p-trunc-len-r 220 \\<br> --o-table table.qza \\<br>
--o-representative-sequences rep-seqs.qza \\<br> --o-denoising-stats
stats.qza \\<br> --verbose]{style="color: #0000FF;"} <br> here --verbose
step will gives messages during the process of denoising.
:::

## View denoised samples stats

::: download
[qiime metadata tabulate \\<br> --m-input-file stats.qza \\<br>
--o-visualization stats.qzv]{style="color: #0000FF;"}
:::

## View feature table

::: download
[qiime feature-table summarize \\<br> --i-table table.qza \\<br>
--m-sample-metadata-file metadata.txt \\<br> --o-visualization
table.qzv]{style="color: #0000FF;"}
:::

## Export feature table

::: download
[qiime tools export \\<br> --input-path rarefied-table.qza \\<br>
--output-path Feature_table/]{style="color: #0000FF;"}
:::

## Convert feature table biom to tsv

::: download
[biom convert \\<br> -i Feature_table/feature-table.biom \\<br> -o
Feature_table/feature-table.tsv \\<br>
--to-tsv\\]{style="color: #0000FF;"}
:::

## View representative sequences

::: download
[qiime feature-table tabulate-seqs \\<br> --i-data rep-seqs.qza \\<br>
--o-visualization rep-seqs.qzv]{style="color: #0000FF;"}
:::

## Alignment

::: download
[qiime phylogeny align-to-tree-mafft-fasttree \\<br> --i-sequences
filtered-rep-seqs.qza \\<br> --output-dir
Aligned/]{style="color: #0000FF;"}
:::

## Diversity

::: download
[qiime diversity core-metrics-phylogenetic \\<br>--i-phylogeny
clustered-sequences/Aligned/rooted_tree.qza \\<br>--i-table
clustered-sequences/rarefied-table.qza \\<br>--p-sampling-depth 27300
\\<br>--m-metadata-file metadata.txt \\<br>--output-dir
clustered-sequences/diversity-rarefied-data]{style="color: #0000FF;"}
:::
